numba/tests/test_np_functions.py

```diff
@@ -2911,6 +2912,12 @@ def input_variations():
             yield np.arange(12).reshape(3, 4)
             yield np.arange(12).reshape(3, 4).T
 
+            # Test cases for `numba.typed.List`
```

Comment:
You'll probably want to add some more types here, maybe float and character? Also, what happens to typed-lists that contain for example other typed-lists or even dictionaries?

Reply:
### Float
Works perfectly!

### Characters
The test fails when the typed list contains characters or `unicode_type`
I presume it is because the `dtype` param of `np.asarray` does not recognize `unicode_type` as a valid datatype.
```python
>>> np.asarray([1, 2], dtype='unicode_type')
TypeError: data type "unicode_type" not understood

>>> np.asarray([1, 2], dtype='unicode')
array(['1', '2'], dtype='<U1')
```

We can do one of two things - 
- add a conditional at https://github.com/numba/numba/pull/5231/files#diff-90991b3b6439443662f04f6aaa03c823R3881 to check if the `dtype` is `unicode_type` and change it to `unicode` - which is (very) hacky. 
- Or, we will need to change the way `numba.types.container.ListType` stores the datatype of unicode lists and change it from `unicode_type` to just `unicode`

### Nested typed-list
If it is a nested typed-list, then we could probably flatten it out. I will add this functionality soon. 

### Dictionaries
For dictionaries, `numba.typed.List` does not support dict elements probably.
```
>>> from numba.typed import List
>>> a = List()
>>> a.append({'a' : 3})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/rohitsanjay/numba/numba/typed/typedlist.py", line 259, in append
    self._initialise_list(item)
  File "/Users/rohitsanjay/numba/numba/typed/typedlist.py", line 224, in _initialise_list
    lsttype = types.ListType(typeof(item))
  File "/Users/rohitsanjay/numba/numba/core/typing/typeof.py", line 33, in typeof
    raise ValueError(msg)
ValueError: cannot determine Numba type of <class 'dict'>
```

Reply:
So, I am not yet sure about what to do in the unicode case, will have to think about it some more. I don't think we can change the way that the typed-list stores unicode. As for the dictionary type, I can only imagine that users might want a record array of `key`  and `value` pairs, but not exactly sure. As for the nested typed-list, I think we would want to convert them to a multidimensional NumPy array. So for example for a singly nested typed-list with typed-lists that contain integers, we would want a 2D NumPy array of integer types.

Reply:
> So for example for a singly nested typed-list with typed-lists that contain integers, we would want a 2D NumPy array of integer types.

Alright, will discuss this over Gitter and I will come up with an implementation soon.